### PR TITLE
Use string dependencies in the get method of the solver [TG-2607]

### PR DESCRIPTION
**Pull request by romainbrenguier**
This is based on https://github.com/diffblue/cbmc/pull/1895 which should be merged soon, so commits up to 123541f can be ignored.
The goal is that `get` uses the information of how a string is obtained with respect to other strings to compute a possible valuation instead of using the model given by the underlying solver.
This can save us several refinement step, as the valuation obtained that way has more chances of satisfying all the string constraints.

### CODE REVIEW

File: src/solvers/refinement/string_refinement_util.cpp [hunk 1]
```diff
@@ -257,7 +257,7 @@ optionalt<exprt> string_insertion_builtin_functiont::eval(
 {
   const auto &input1_value = eval_string(input1, get_value);
   const auto &input2_value = eval_string(input2, get_value);
-  if(!input2_value.has_value() || !input1_value.has_value())
+  if(!input1_value.has_value() || !input2_value.has_value())
```
Comment (tautschnig) [inline src/solvers/refinement/string_refinement_util.cpp]: The change seems to be completely unrelated to the commit message.

File: src/solvers/refinement/string_refinement.cpp [hunk 2]
```diff
@@ -757,6 +757,7 @@ decision_proceduret::resultt string_refinementt::dec_solve()
 
   // Initial try without index set
   const auto get = [this](const exprt &expr) { return this->get(expr); };
+  dependences.clean_cache();
```
Comment (tautschnig) [inline src/solvers/refinement/string_refinement.cpp]: I believe the plural is "dependencies". Notably, your code sometimes already uses "dependencies".

File: regression/jbmc-strings/StringConcat/Test.java [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+	sb.append(' ');
+	sb.append('|');
+	sb.append('\n');
+
+	assert(false);
+	return sb.toString();
+    }
+
+    public String bufferNonDetLoop2(int width, String c) {
+	if(width < 5 || c.length() > width)
```
Comment (tautschnig) [inline regression/jbmc-strings/StringConcat/Test.java]: Strange formatting.

File: regression/jbmc-strings/StringConcat/test_buffer_det.desc [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+Test.class
+--refine-strings --string-max-length 1000 --string-max-input-length 100 --function Test.bufferDet --depth 10000  --verbosity 10
+^EXIT=10$
+^SIGNAL=0$
+^VERIFICATION FAILED$
+--
+check_SAT: got SAT but the model is not correct
+--
+Using the string dependences information to compute the model, refinement
+should not be needed when the execution is deterministic.
```
Comment (tautschnig) [inline regression/jbmc-strings/StringConcat/test_buffer_det.desc]: No end-of-line at end-of-file. (Applies elsewhere as well.)

File: src/solvers/refinement/string_refinement_util.h [hunk 5]
```diff
@@ -140,4 +141,203 @@ class equation_symbol_mappingt
   std::unordered_map<std::size_t, std::vector<exprt>> strings_in_equation;
 };
 
+/// Base class for string functions that are built in the solver
+class string_builtin_functiont
```
Comment (LAJW) [inline src/solvers/refinement/string_refinement_util.h]: When creating a virtual class hierarchy, the base class destructor should be defined and marked as virtual. i.e.

```
virtual ~string_builtin_functiont() = default;
```

Otherwise, you'll end up with memory leaks.

File: src/solvers/refinement/string_refinement_util.h [hunk 6]
```diff
@@ -140,4 +141,203 @@ class equation_symbol_mappingt
...
+{
+public:
+  string_builtin_functiont() = default;
```
Comment (LAJW) [inline src/solvers/refinement/string_refinement_util.h]: If this is not intended to be instantiated, you could mark this constructor as `protected`.

File: src/solvers/refinement/string_refinement_util.h [hunk 7]
```diff
@@ -140,4 +141,203 @@ class equation_symbol_mappingt
...
+  std::vector<array_string_exprt> string_arguments() const override
+  {
+    return under_test;
+  }
+};
+
+/// Keep track of dependencies between strings.
+/// Each string points to builtin_function calls on which it depends,
+/// each builtin_function points to the strings on which the result depend.
+class string_dependencest
```
Comment (LAJW) [inline src/solvers/refinement/string_refinement_util.h]: Typo - `dependencies` - missing an `i`. Also, documentation could use better grammar.

File: src/solvers/refinement/string_refinement.cpp (hunk 2 above)
Comment (romainbrenguier) [reply src/solvers/refinement/string_refinement.cpp]: my understanding is that dependences is the the plural of dependence and dependencies the plural of dependency. Anyway I'd prefer to keep only one...

**Comment by tautschnig** [comment]
I'd appreciate if a native speaker could chime in, but my reading of https://en.wiktionary.org/wiki/dependence#English vs. https://en.wiktionary.org/wiki/dependency#English is that the latter is the appropriate technical term.

**Comment by martin-cs** [comment]
@tautschnig I would use depend / dependency / dependencies (as is "x == f(x) depends on x and all of f's dependencies") in technical computing English.  Will review code when I've had a chance to look at the previous PR.

File: src/solvers/refinement/string_refinement.cpp [hunk 8]
```diff
@@ -680,7 +680,7 @@ decision_proceduret::resultt string_refinementt::dec_solve()
 #ifdef DEBUG
   output_equations(debug(), equations, ns);
 
-  string_dependencest dependences;
+  string_dependenciest dependences;
```
Comment (peterschrammel) [inline src/solvers/refinement/string_refinement.cpp]: Can you also rename the variable, please?

File: src/solvers/refinement/string_refinement_util.cpp [hunk 9]
```diff
@@ -172,6 +172,39 @@ equation_symbol_mappingt::find_equations(const exprt &expr)
...
+  const auto &input_value = eval_string(input, get_value);
+  if(!input_value.has_value())
+    return {};
+
+  std::vector<mp_integer> arg_values;
+  const auto &insert = std::back_inserter(arg_values);
+  std::transform(args.begin(), args.end(), insert, [&](const exprt &e) { // NOLINT
+    if(const auto val = numeric_cast<mp_integer>(get_value(e)))
+      return *val;
+    // INVARIANT(get_value(e).id() == ID_unknown);
```
Comment (peterschrammel) [inline src/solvers/refinement/string_refinement_util.cpp]: Remove commented out code

File: src/solvers/refinement/string_refinement_util.cpp [hunk 10]
```diff
@@ -526,7 +521,12 @@ bool add_node(
   if(
     const auto &string_result =
       dependencies.get_builtin_function(builtin_function_node).string_result())
+  {
     dependencies.add_dependency(*string_result, builtin_function_node);
+    auto &node = dependencies.get_node(*string_result);
+    node.result_from = builtin_function_node;
+
```
Comment (peterschrammel) [inline src/solvers/refinement/string_refinement_util.cpp]: remove spurious blank line

**Review by peterschrammel — COMMENTED** [review]
Looks good. There are failing tests, though.

**Review by peterschrammel — APPROVED** [review]

**Review by LAJW — APPROVED** [review]
